pkg/manifest/output.go

```diff
@@ -24,10 +24,11 @@ func (m *Manifest) ToTarGZ() (io.Reader, error) {
 			return nil, err
 		}
 
-		bytes, err = decryptSOPS(bytes)
-		if err != nil {
-			return nil, err
-		}
+		// Remove to allow downstream to handle
```

Comment:
I think this would be better suited as a flag in the CRD.
CC: @StrongMonkey @kinarashah 

Perhaps under `BundleDeploymentOptions`, `DecryptSOPS` as `*bool` set to `true` in our defaults?
https://github.com/rancher/fleet/blob/master/pkg/apis/fleet.cattle.io/v1alpha1/bundle.go#L177

Reply:
hi @nickgerace , I added your recommendations, tested and confirmed it works locally but it's failing the tests here on github, not sure why (no access)

Reply:
No problem @mumblez. Thank you for your quick changes. Before taking a look at the CI, I think there are some additional places we should look at. I'll make comments at those places.

Reply:
sure, np

Reply:
@nickgerace just wondering if you know where the sops decryption is happening by default, is it on the controller (`fleet` chart) or on the agent cluster (`fleet-agent`), or is it `fleet-agent` on the controller server?

Reply:
Hi @nickgerace, not sure if any comments were left, might be some permission issue.
(Note, on the side I'm trying to get sops to work correctly, I managed to self contain the `output.go` and can confirm passing in the right credentials and env vars works, but doing the same to the `fleet` deployment container makes no difference, maybe there's some other job / pod doing it but can't figure it out :( )

Reply:
Hi @mumblez. Do not worry; still need to re-review. Been a busy week :)

Reply:
np

Reply:
At a cursory glance, functions/methods that call `Deploy --> ToChart --> ToTarGZ` from `pkg/helmdeployer/deployer.go` [here](https://github.com/rancher/fleet/blob/master/pkg/helmdeployer/deployer.go#L150) are what we need to look out for. I think putting the field in `BundleDeploymentOptions` is a good place to start.

This method gets called twice:
1. The `agent` deployer will call this method once in `modules/agent/pkg/deployer/manager.go` [here](https://github.com/rancher/fleet/blob/f8008e0437b7c77c08e9a3e840e3d19142fa6537/modules/agent/pkg/deployer/manager.go#L113). I believe this answers the `agent` part of your question.
2. The `helmdeployer` template will call this method once in `pkg/helmdeployer/template.go` [here](https://github.com/rancher/fleet/blob/cfbe1dce4e6a5d836ff949b2eee33dbde4da79ec/pkg/helmdeployer/template.go#L17). **However,** this is where we need to make sure we understand why/when this is called. Let's continue below...

Following the `helmdeployer` path, we trace to this unit test.
We can optionally add the field here. I do not think it is necessary though.
```modules/cli/match/match.go <-- modules/cli/cmds/test.go```

The other trace leads us down this path, which is where we need to look.
```pkg/controllers/bundle/controller.go```
Within this controller, we have where the options are calculated [here](https://github.com/rancher/fleet/blob/ffa88d184cab87c4a53c17c65df45c0d213358c3/pkg/controllers/bundle/controller.go#L156).
Keeping this in mind, we have two ways to go from here.

1. Add a field to the `GitRepo` CRD since that is our user-facing CRD. `Bundle` is an internal unit. We can essentially at a similar `*bool` field, and set to `true` if it is unset.
2. Add value to the Fleet chart(s) for this option. I prefer the first option since affecting the entire chart seems to be unnecessary. The benefit to this approach is that you have a "global" state on how to handle `sops`.

I hope this mini-rant helps :)
What do you think? I'm thinking we should go with option 1 above.

CC: @StrongMonkey @kinarashah

Reply:
@nickgerace thank you, that makes perfect sense, setting in `GitRepo` is a cleaner solution, I will try to refactor taking your points onboard, will keep you posted.

Reply:
Hi @nickgerace 

I think it might be better if I try to pass in the sops credentials via helm vs adding this `decrypt` var to numerous places, in theory there won't be any code or logic changes as by default it will try to decrypt only if `sops` related resources are found.

I assume it's the fleet controller (management server) that is doing the decryption, but when I amend it's helm chart to mount the secret and expose the environment variable it makes no difference, after fleet controller extension there is also fleet agent on the management server, I assume it is this that is doing the decryption?, is the controller installing the agent itself and if so how would you recommend I pass down the sops credentials?

Or is it the downstream fleet agent that is doing the decryption?

Many Thanks

Reply:
>  in theory there won't be any code or logic changes as by default it will try to decrypt only if sops related resources are found

I could be wrong here, but isn't this the current behavior?

```go
func decryptSOPS(data []byte) ([]byte, error) {
	if !bytes.Contains(data, []byte("sops:")) {
		return data, nil
	}
        // --snip--
}
```

As far as your idea goes, we would only need `decrypt` in one place, and have it defined in two CRDs, which I don't think is big overhead. However, I'll answer your questions to help us confirm that your second idea might be a better one :) 

> I assume it's the fleet controller (management server) that is doing the decryption, but when I amend it's helm chart to mount the secret and expose the environment variable it makes no difference, after fleet controller installation there is also fleet agent on the management server (installed by fleet controller I assume), I assume it is this that is doing the decryption?

I believe it is the `fleet-agent` from my previous trace of the codebase. Just curious, for you "allow(ing) downstream to handle" sops, does it matter which pod we are concerned with? I'm curious to the motivation behind this question.

> is the controller installing the agent itself and if so how would you recommend I pass down the sops credentials?
> Or is it the downstream fleet agent that is doing the decryption?

Admittedly, I am unsure. I can get back to you with an exact answer here. Regardless of which agent performs sops decryption, if you're going with this second approach (assuming we still use `GitRepo` as the payload), then we would most likely store the credentials as a `Secret` that eventually becomes a field in the `Bundle` or `BundleDeployment` (or other) CRD.

Reply:
Alright, I spoke to @StrongMonkey and while your new idea is really good and clean, I think we should go with the original approach of adding a `*bool` to GitRepo. The new approach of passing SOPS credentials via Helm is great, but we are evaluating how we do SOPS decryption overall, and I'd prefer to not overload this PR and send you in multiple directions. Adding the bool field/flag should be sufficient for now, I believe. Does that sound good to you? @mumblez

Reply:
yes, this is the current behaviour (if `sops:` is detected in the resource)

> I believe it is the fleet-agent from my previous trace of the codebase. Just curious, for you "allow(ing) downstream to handle" sops, does it matter which pod we are concerned with? I'm curious to the motivation behind this question.

Am just trying to work out which container is doing the decryption so I can test and verify mounting the credentials and exporting the env var works. (Preferably this happens on the controller vs downstream), then I can work out how to amend the helm charts and configs to supply this. 

Currently we can helm deploy the `fleet-controller` and `fleet-agent` and configure their helm charts, but when `fleet-controller` also sets it is local `fleet-agent`; I'm unsure how to feed it the sops credentials, ideally we can just amend the helm chart / k8s manifest but we can't do that for `fleet-agent` on the control server.

Reply:
> Alright, I spoke to @StrongMonkey and while your new idea is really good and clean, I think we should go with the original approach of adding a *bool to BundleDeploymentOptions. The new approach of passing SOPS credentials via Helm is great, but we are evaluating how we do SOPS decryption overall, and I'd prefer to not overload this PR and send you in multiple directions. Adding the bool field/flag should be sufficient for now, I believe. Does that sound good to you? @mumblez

Sure, np, I'm just a little stuck on how to access from the gitrepo config, e.g. I can add the value to the `GitRepoSpec` struct in `pkg/apis/fleet.cattle.io/v1alpha1/git.go` but not sure how to access from the functions / bundles, any help appreciated.

Reply:
Actually, you were correct the first time and I led you astray by accident @mumblez. Apologies for that; long week with release.

The `BundleDeploymentOptions` should be configurable via `fleet.yaml`, like in the docs file you have above. We should not add anything to `GitRepo`, like I initially suggested. Apologies for that.

Reply:
Also, I see what you mean. I believe both the `fleet-controller` and the _local_ `fleet-agent` perform decryption. I am confirming whether or not the downstream ones do as well, but I believe that _not_ to be the case.